calico.go

```diff
@@ -44,80 +45,113 @@ func init() {
 	// Unmarshal the network config, and perform validation
-	conf := NetConf{}
+	conf := types.NetConf{}
 	if err := json.Unmarshal(args.StdinData, &conf); err != nil {
 		return fmt.Errorf("failed to load netconf: %v", err)
 	}
 
-	cniVersion := conf.CNIVersion
+	utils.ConfigureLogging(conf.LogLevel)
 
-	ConfigureLogging(conf.LogLevel)
+	// Allow the nodename to be overridden by the network config
+	nodename = utils.DetermineNodename(conf)
 
-	workload, orchestrator, err := GetIdentifiers(args)
+	// Extract WEP identifiers such as pod name, pod namespace (for k8s), containerID, workload name.
+	epIDs, err := utils.GetIdentifiers(args, nodename)
 	if err != nil {
 		return err
 	}
 
-	logger := CreateContextLogger(workload)
+	log.WithField("EndpointIDs", epIDs).Info("Extracted identifiers")
+	log.WithField("NetConfg", conf).Info("Loaded CNI NetConf")
 
-	// Allow the nodename to be overridden by the network config
-	updateNodename(conf, logger)
+	calicoClient, err := utils.CreateClient(conf)
+	if err != nil {
+		return err
+	}
 
-	logger.WithFields(log.Fields{
-		"Orchestrator": orchestrator,
-		"Node":         nodename,
-		"Workload":     workload,
-		"ContainerID":  args.ContainerID,
-	}).Info("Extracted identifiers")
+	// Remove the endpoint from the epIDs so we can get a WEP name prefix.
+	epIDs.Endpoint = ""
 
-	logger.WithFields(log.Fields{"NetConfg": conf}).Info("Loaded CNI NetConf")
-	calicoClient, err := CreateClient(conf)
+	// Calculate the workload name from the WEP specific identifiers
+	// for the given orchestrator.
+	wepPrefix, err := epIDs.CalculateWorkloadEndpointName(true)
 	if err != nil {
-		return err
+		return fmt.Errorf("error constructing WorkloadEndpoint prefix: %s", err)
 	}
 
-	// Always check if there's an existing endpoint.
-	endpoints, err := calicoClient.WorkloadEndpoints().List(api.WorkloadEndpointMetadata{
-		Node:         nodename,
-		Orchestrator: orchestrator,
-		Workload:     workload})
+	ctx := context.Background()
+
+	// Check if there's an existing endpoint.
+	endpoints, err := calicoClient.WorkloadEndpoints().List(ctx, options.ListOptions{Name: wepPrefix, Namespace: epIDs.Namespace, Prefix: true})
 	if err != nil {
 		return err
 	}
 
-	logger.Debugf("Retrieved endpoints: %v", endpoints)
+	var logger *log.Entry
+	if epIDs.Orchestrator == "k8s" {
+		logger = log.WithFields(log.Fields{
+			"WorkloadEndpoint": fmt.Sprintf("%s%s", wepPrefix, epIDs.Endpoint),
+			"ContainerID":      epIDs.ContainerID,
+			"Pod":              epIDs.Pod,
+			"Namespace":        epIDs.Namespace,
+		})
+	} else {
+		logger = log.WithFields(log.Fields{
+			"ContainerID": epIDs.ContainerID,
+		})
+	}
+
+	logger.Debugf("Retrieved list of endpoints: %v", endpoints)
 
 	var endpoint *api.WorkloadEndpoint
-	if len(endpoints.Items) == 1 {
-		endpoint = &endpoints.Items[0]
+	// Put the endpoint name back in the identifiers
+	// so we can get the exact WEP match from the WEP list.
+	epIDs.Endpoint = args.IfName
```

Comment:
Logic here is still not quite right.

You are setting the endpoint back to the original value too early - which means you are checking to see if the enumerated endpoint also has the same endpoint ID.  In which case, you might just as well have done an exact get rather than a prefix list.

Your logic is:
e.g.
WEP ids are `node1, cni, container1, eth0`

You do a prefix query by removing the `eth0` - so you get get a list of the endpoints for container1.
You get returned an endpoint for eth1.
You drop it on the floor because you are checking again for eth0.

Reply:
You need to check it's a valid endpoint *before* putting the endpoint back in.

You then need to extract the endpoint ID from the workload Spec.

Reply:
A test for this case would have probably highlighted this early on and saved some cycles.